### PR TITLE
Add persistent state directory to python agents

### PR DESCRIPTION
**Pull request by nicoloboschi**
Changes:
- Added new Python AgentContext class to the API
- An agent can access it by defining
```
from langstream import SimpleRecord, Processor, AgentContext
import logging, os


class Exclamation(Processor):
    def init(self, config, context: AgentContext):
        self.context = context
    def process(self, record):
        directory = self.context.get_persistent_state_directory() 
``` 

if the agent declares init with only argument, it works as before
- Added end to end test

### CODE REVIEW

File: langstream-runtime/langstream-runtime-impl/src/main/python/Pipfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-[packages]
+    [packages]
```
Comment (eolivelli) [inline langstream-runtime/langstream-runtime-impl/src/main/python/Pipfile:1]: is this intended ?

File: langstream-runtime/langstream-runtime-impl/src/main/python/langstream/api.py [hunk 2]
```diff
@@ -60,11 +61,25 @@ def headers(self) -> List[Tuple[str, Any]]:
...
+        """Return a path pointing to the stateful agent directory. Return None if not configured in the agent."""
+        pass
+
+
 
 class Agent(ABC):
     """The Agent interface"""
 
-    def init(self, config: Dict[str, Any]):
+    def init(self, config: Dict[str, Any], context: AgentContext):
```
Comment (eolivelli) [inline langstream-runtime/langstream-runtime-impl/src/main/python/langstream/api.py:82]: is this a breaking change ?

maybe we should have same approach as in Java and have a "setAgentContext" function

File: langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/__main__.py [hunk 3]
```diff
@@ -27,12 +27,15 @@
...
 
-    if len(sys.argv) != 3:
+    if len(sys.argv) <= 3:
         print("Missing gRPC target and python class name")
         print("usage: python -m langstream_grpc <target> <config>")
         sys.exit(1)
 
-    server = AgentServer(sys.argv[1], sys.argv[2])
+    context_config = {}
+    if len(sys.argv) > 3:
```
Comment (eolivelli) [inline langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/__main__.py:36]: the process is always launched by the same version of the java runtime, there is no need to do this

**Review by eolivelli — CHANGES_REQUESTED** [review]
I have left some feedback.

@cbornet please review as well

File: langstream-runtime/langstream-runtime-impl/src/main/python/langstream/api.py (hunk 2 above)
Comment (nicoloboschi) [reply langstream-runtime/langstream-runtime-impl/src/main/python/langstream/api.py:82]: no it is not, I added a test for the old syntax

File: langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/__main__.py (hunk 3 above)
Comment (nicoloboschi) [reply langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/__main__.py:36]: 👍

File: langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/api.py [hunk 4]
```diff
@@ -61,10 +62,22 @@ def headers(self) -> List[Tuple[str, Any]]:
...
 
 
+class AgentContext(ABC):
+    """The Agent context interface"""
+
+    def __init__(self):
+        """Initialize the agent context."""
+        pass
+
+    def get_persistent_state_directory(self):
```
Comment (cbornet) [inline langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/api.py:72]: abstractmethod ?

File: langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/grpc_service.py [hunk 5]
```diff
@@ -324,27 +319,44 @@ def to_grpc_value(self, value) -> Tuple[Optional[Schema], Optional[Value]]:
...
     module_name = full_class_name[: -len(class_name) - 1]
     module = importlib.import_module(module_name)
     agent = getattr(module, class_name)()
-    call_method_if_exists(agent, "init", configuration)
+    context_impl = AgentContextImpl(configuration, context)
+    call_method_if_exists(agent, "init", configuration, context_impl)
     return agent
 
 
+class AgentContextImpl(AgentContext):
```
Comment (cbornet) [inline langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/grpc_service.py:341]: rename DefaultAgentContext ?

File: langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/api.py (hunk 4 above)
Comment (cbornet) [inline langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/api.py:72]:
```suggestion
    @abstractmethod
    def get_persistent_state_directory(self):
```

**Comment by cbornet** [comment]
LGTM. Just some minor comments.
Congrats on the Python PR !

File: langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/grpc_service.py (hunk 5 above)
Comment (nicoloboschi) [reply langstream-runtime/langstream-runtime-impl/src/main/python/langstream_grpc/grpc_service.py]: done

**Comment by nicoloboschi** [comment]
> LGTM. Just some minor comments. Congrats on the Python PR !

Thanks I addressed your comments

**Review by eolivelli — APPROVED** [review]

**Pull request merged**